충남대 BE_좌진우 1단계 - 카카오 로그인

jinwoojwa opened the pull request:
## Step 1 - 카카오 로그인

✅ 구현 내용

- 카카오 로그인 인가 코드 받기 구현 (/auth/kakao/login)
- 인가 코드 수신 및 토큰 요청 처리 (/auth/kakao/callback)
- 외부 API와의 통신을 위해 `RestClient` 사용
- `application.properties`에 카카오 클라이언트 ID와 리다이렉트 URI 설정 추가
- 토큰 응답을 DTO로 매핑하여 향후 로직에 활용할 수 있도록 구현

💻 테스트

- 테스트 URL (로그인 시작): http://localhost:8080/auth/kakao/login


- 액세스 토큰 발급 확인

<img width="627" height="119" alt="image" src="https://github.com/user-attachments/assets/c585ae93-93f1-414f-bcc7-de49fd833ccd" />

jinwoojwa committed:
docs(README): step1 기능 요구 사항 및 구현 사항 정리

jinwoojwa committed:
feat: 카카오 로그인 인가 코드 및 토큰 발급 기능 구현

leaf-upper commented on README.md:
```diff
@@ -16,4 +17,53 @@
+
+<br>
+
+## Step 1 - 카카오 로그인
```
👍 👍 리드미를 잘 정리해주셨네요 고생하셨습니다 ㅎㅎ

leaf-upper commented on src/main/java/gift/auth/controller/KakaoAuthController.java:
```diff
@@ -0,0 +1,58 @@
+    private String redirectUri;
+
+    @GetMapping("/login")
+    public void redirectToKakaoAuth(HttpServletResponse response) throws IOException {
```
로그인 페이지로 오면 "카카오 로그인" 페이지로 redirect 하는것 같아요.
clientId와 redirectUri는 Property값으로 외부에서 받는것 같습니다.

1. property는 별도 클래스 혹은 record를 선언하여 관리해보는건 어떨까요? (외부에서 주입받고 변하지 않는 값일것 같아서 final 키워드 혹은 record로 관리하시면 좋을것 같아요)
2. kakaoAuthUrl도 property로 관리할 수 있지 않을까요?

leaf-upper commented on src/main/java/gift/auth/controller/KakaoAuthController.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    @GetMapping("/callback")
+    public ResponseEntity<KakaoTokenResponseDto> kakaoCallback(@RequestParam("code") String code) {
```
code를 통해서 카카오로부터 액세스 토큰을 받아오신것 같아요.
해당 토큰을 통해 유저의 정보를 입력받고 이를 기반으로 로그인을 실제로 하셔야
이번 스텝을 마무리할 수 있을것 같아요.
추가적인 구현 부탁드립니다.

leaf-upper commented on src/main/java/gift/auth/dto/KakaoTokenResponseDto.java:
```diff
@@ -0,0 +1,9 @@
+package gift.auth.dto;
+
+public record KakaoTokenResponseDto(
+        String token_type,
```
Java는 주로 카멜케이스를 사용해요. 카멜케이스로 필드명을 관리하시면 좋을것 같아요.

leaf-upper commented on src/main/java/gift/auth/controller/KakaoAuthController.java:
```diff
@@ -0,0 +1,58 @@
+                "&redirect_uri=" + redirectUri +
+                "&code=" + code;
+
+        ResponseEntity<KakaoTokenResponseDto> response = restClient.post()
```
네트워크 통신중 오류가 발생하면 어떻게 할까요?

leaf-upper reviewed (changes requested):
안녕하세요 진우님 1단계 고생하셨습니다.
몇가지 코멘트를 남겨드렸으니 시간되실때 확인후에
추가적인 구현 및 수정 부탁드려요~
고생하셨습니다.

jinwoojwa committed:
fix: KakaoTokenResponseDto 필드명 스네이크 케이스에서 카멜 케이스로 변경

jinwoojwa committed:
refactor: 카카오 OAuth 설정을 record 기반 프로퍼티 클래스로 분리 및 관리

- application.properties에 kakao 관련 설정 추가
- KakaoOauthProperties record 생성하여 설정값 바인딩 처리
- @EnableConfigurationProperties를 위한 별도 Config 클래스 생성
- KakaoAuthController에서 설정값 직접 참조 대신 프로퍼티 클래스를 사용하도록 리팩토링

jinwoojwa committed:
feat: Kakao OAuth API 호출에 대한 예외 처리 및 글로벌 예외 핸들러 추가

- RestClient 호출 시 onStatus를 활용해 4xx, 5xx 응답에 대해 커스텀 예외 던지기 구현
- KakaoOAuthClientException, KakaoOAuthServerException 커스텀 예외 클래스 추가
- GlobalExceptionHandler에 두 예외 처리 핸들러 추가

jinwoojwa commented:
🧾 변경 사항

1️⃣ KakaoTokenResponseDto 필드명 변경

- 스네이크 케이스에서 카멜 케이스로 변경했습니다.

2️⃣ 로그인 설정 값 분리

- 로그인 관련 설정 값들을  KakaoOauthProperties로 분리하여 처리했습니다.
- 컨트롤러에서도 직접 참조하는 대신 프로퍼티 클래스를 이용하도록 했습니다.

3️⃣ API 호출 시 발생할 수 있는 예외 처리 추가

- RestClient 호출 시 발생할 수 있는 4xx, 5xx 에러를 처리하는 커스텀 예외를 구현했습니다.
- 구현한 예외를 처리하는 핸들러도 추가했습니다.

❓ 질문 사항

로그인을 구현하려다 궁금한 사항이 생겨 질문드립니다.

다양한 애플리케이션을 사용하다 보면, 해당 앱에 직접 회원가입해서 로그인하는 경우와 카카오 같은 OAuth를 통해 로그인하는 경우가 별개의 계정으로 처리되는 것으로 보입니다.
또한, 이전까지의 미션에서 회원 가입, 로그인을 하기 위해 사용했던 Member 엔티티는 일반 회원가입(이메일/비밀번호 기반)을 전제로 설계되었다고 생각합니다.
그렇다면 OAuth 로그인을 구현할 때는 기존 Member 엔티티와는 별도로 소셜 로그인을 위한 새로운 회원 엔티티를 만들어야 할까요?
혹은 하나의 Member 엔티티 안에서 소셜 로그인 방식까지 함께 처리해야 하는지 궁금합니다.

++ 추가로 실제 애플리케이션에서는 소셜 로그인과 일반 로그인을 분리하여 구현하는지 궁금합니다.

jinwoojwa committed:
fix: KakaoTokenResponseDto에 @JsonProperty 적용

jinwoojwa committed:
chore: API 키를 secrets.properties로 분리하고 .gitignore에 추가

jinwoojwa committed:
feat: KakaoOauthProperties에 tokenUrl, userInfoUrl 프로퍼티 추가 및 매핑

jinwoojwa committed:
feat: 카카오 로그인 기능 추가 및 리팩토링

- KakaoAuthController, KakaoAuthService, KakaoOAuthClient로 역할 분리
- 사용자 정보 처리 DTO 추가

jinwoojwa commented:
생각해보니 Member를 그대로 사용해도 괜찮을 것 같아 로그인 구현까지 완료했습니다.
이후 미션 진행하면서 추가로 로그인 페이지와 관련 컨트롤러도 구현해 볼 예정입니다.
변경 사항은 아래와 같습니다.

🧾 변경 사항

1️⃣ DTO에 `@JsonProperty` 적용

- key를 매핑시켜주기 위해 `@JsonProperty`를 적용했습니다.

2️⃣ API키 관리

- API키를 secrets.properties 파일에서 관리하고, `.gitignore`로 지정해주었습니다.

3️⃣ 로그인 기능 추가 및 구조 리팩토링

- 로그인 기능을 추가하고, 기존의 컨트롤러 코드를 컨트롤러, 서비스, 클라이언트 코드로 분리했습니다.

leaf-upper commented on README.md:
```diff
@@ -16,4 +17,53 @@
+<br>
+
+## Step 1 - 카카오 로그인
+
```
질문 주신 부분에 대해서 여기서 답변드릴게요~

```
다양한 애플리케이션을 사용하다 보면, 해당 앱에 직접 회원가입해서 로그인하는 경우와 카카오 같은 OAuth를 통해 로그인하는 경우가 별개의 계정으로 처리되는 것으로 보입니다.
또한, 이전까지의 미션에서 회원 가입, 로그인을 하기 위해 사용했던 Member 엔티티는 일반 회원가입(이메일/비밀번호 기반)을 전제로 설계되었다고 생각합니다.
그렇다면 OAuth 로그인을 구현할 때는 기존 Member 엔티티와는 별도로 소셜 로그인을 위한 새로운 회원 엔티티를 만들어야 할까요?
혹은 하나의 Member 엔티티 안에서 소셜 로그인 방식까지 함께 처리해야 하는지 궁금합니다.

++ 추가로 실제 애플리케이션에서는 소셜 로그인과 일반 로그인을 분리하여 구현하는지 궁금합니다.
```
저는 OAuth와 아이디  패스워드를 사용하여 로그인 하는 계정을 다르게 하는 경우는 거의 보지 못한것 같아요.
같은 회원에 대해서 로그인 하는 방식이 다른것이지, 계정이 달라지는게 아닙니다. Member는 특정 유저에 대해서 식별하고 있고, 해당 유저가 OAuth를 통해 본인을 식별하든, 아이디 패스워드를 통해 식별하든 서비스에서는 해당 유저의 신원을 식별할 수 있다면 상관없습니다. 인증별로 Member의 엔티티를 구별해서 둘 필요는 없을것 같아요.

leaf-upper commented on src/main/java/gift/auth/service/DefaultKakaoAuthService.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Override
+    @Transactional
+    public MemberResponseDto loginWithKakao(String code) {
```
카카오로 로그인하고나서, 이후 유저가 로그인 한 상태로 API를 요청할수 있도록 클라이언트쪽으로 본인을 식별할 수 있는 토큰값을 내려드리는게 좋지 않을까요?

leaf-upper commented on src/main/resources/application.properties:
```diff
@@ -19,3 +19,11 @@ spring.jpa.hibernate.ddl-auto=create
+
+spring.config.import=optional:classpath:secrets.properties
+
+kakao.client-id=${kakao.api.key}
```
👍 👍 
application 프로퍼티 잘정리해주셨습니다.

leaf-upper commented on src/main/java/gift/auth/client/KakaoOAuthClient.java:
```diff
@@ -0,0 +1,58 @@
+        return restClient.post()
+                .uri(kakaoProps.tokenUrl())
+                .contentType(MediaType.APPLICATION_FORM_URLENCODED)
+                .body(body)
```
👍 👍 예외처리 잘해주셨습니다. 네트워크 통신중에는 오류가 발생할 수 있다는 사실에 대해서 기억하면서 개발하면 좋을것 같아요. 특정 상황에서는 소켓 연결도 안되는 상황이 발생할 수 있습니다.

leaf-upper reviewed (approved):
안녕하세요 진우님 1단계 고생 많으셨습니다.
이전에 코멘트 드린 부분은 거의 수정된것 같아서
다음단계 진행해주셔도 좋을것 같아요~
몇가지 부족한 부분은 코멘트 남겨드렸습니다. 다음단계에서
수정이 필요한 부분은 같이 수정해서 PR주세요~
